openwrt/luci
# luci-app-example: Update with more documentation, more examples

Summary
=======

The existing luci-app-example is very basic, and provides very little context to someone just getting started with LuCI development.

The changes in this PR are basically "I wish I'd had this information and context available to me", as I've recently worked on a PR to provide a UI for AdGuard Home, and it took me several days of hunting around in the wiki, other LuCI apps, and general 'net searches.

Documentation changes:

* add a structure.md file to explain the structure of a LuCI app
* add a BUILDING.md file to provide a quick overview of how to build the .ipk file for the app
* add a YAML.md to detail working with YAML files
* update README.md with some details about the Javascript files in the example
* update README.md with a note about ACLs
* update README.md with a note about translations

ACL changes:

* update the ACL file based on testing, removing unused grants

Code changes:

* update the menu file with links to the new Javascript UI files
* add RPC examples for a dedicated ubus RPC endpoint that's not UCI
* add a Lua script to answer RPC requests via rpcd

The Lua script demonstrates a few techniques, such as

* converting Lua tables to JSON
* argument parsing, with explanation about why it's done this way (luci.ddns did not explain the logic, nor did the wiki)
* how to test the Lua script without ubus
* how to test the Lua script with ubus

Three variants of the rpc.js file are provided. The first one is just calling RPC endpoints and rendering what could be equated to a status page; editing is turned off. The second and third call the same RPC endpoint, and demonstrate two different layout approaches - TableSection vs TypedSection.

Notes
=====

This is a re-spin of https://github.com/openwrt/luci/pull/5857, due to a ton of conflicts when trying to rebase my original PR via git commands. @jow- 's feedback about the YAML code has been taken on board, and split out to documentation that someone has to manually apply.

Testing
=======

All testing has been manual.

ACLs were refined by moving acl.d/luci-compat.json out of the way to see what broke in the UI, and then tweaking the ACLs until the UI worked again.

Arg parsing for the Lua script is based on the luci.ddns script, and it feels like the parseInput and validateArgs functions should be a core library so that Lua scripts for rpcd have a unified way of working.

Signed-off-by: Duncan Hill <github.com@cricalix.net>
